homeassistant/components/fan/dyson.py

```diff
@@ -238,6 +324,252 @@ def supported_features(self) -> int:
+                return SPEED_MEDIUM
+            if self._device.state.speed in low_speeds:
+                return SPEED_LOW
+            if self._device.state.speed in medium_speeds:
+                return SPEED_MEDIUM
+            if self._device.state.speed in high_speeds:
+                return SPEED_HIGH
+        return None
+
+    @property
+    def dyson_speed(self):
+        """Return the current speed."""
+        from libpurecool.const import FanSpeed
+
+        if self._device.state:
+            if self._device.state.speed == FanSpeed.FAN_SPEED_AUTO.value:
+                return self._device.state.speed
+            return int(self._device.state.speed)
+        return None
+
+    @property
+    def night_mode(self):
+        """Return Night mode."""
+        return self._device.state.night_mode == "ON"
+
+    @property
+    def auto_mode(self):
+        """Return Auto mode."""
+        return self._device.state.auto_mode == "ON"
+
+    @property
+    def angle_low(self):
+        """Return angle high."""
+        return self._device.state.oscillation_angle_low
+
+    @property
+    def angle_high(self):
+        """Return angle low."""
+        return self._device.state.oscillation_angle_high
+
+    @property
+    def flow_direction_front(self):
+        """Return frontal flow direction."""
+        return self._device.state.front_direction == 'ON'
+
+    @property
+    def timer(self):
+        """Return timer."""
+        return self._device.state.sleep_timer
+
+    @property
+    def speed_list(self) -> list:
+        """Get the list of available speeds."""
+        return [SPEED_LOW, SPEED_MEDIUM, SPEED_HIGH]
+
+    @property
+    def dyson_speed_list(self) -> list:
+        """Get the list of available dyson speeds."""
+        from libpurecool.const import FanSpeed
+
+        supported_speeds = [
+            int(FanSpeed.FAN_SPEED_1.value),
+            int(FanSpeed.FAN_SPEED_2.value),
+            int(FanSpeed.FAN_SPEED_3.value),
+            int(FanSpeed.FAN_SPEED_4.value),
+            int(FanSpeed.FAN_SPEED_5.value),
+            int(FanSpeed.FAN_SPEED_6.value),
+            int(FanSpeed.FAN_SPEED_7.value),
+            int(FanSpeed.FAN_SPEED_8.value),
+            int(FanSpeed.FAN_SPEED_9.value),
+            int(FanSpeed.FAN_SPEED_10.value),
+        ]
+
+        return supported_speeds
+
+    @property
+    def supported_features(self) -> int:
+        """Flag supported features."""
+        return SUPPORT_OSCILLATE | \
+            SUPPORT_SET_SPEED
+
+    @property
+    def device_state_attributes(self) -> dict:
+        """Return optional state attributes."""
+        return {
+            ATTR_NIGHT_MODE: self.night_mode,
+            ATTR_AUTO_MODE: self.auto_mode,
+            ATTR_ANGLE_LOW: self.angle_low,
+            ATTR_ANGLE_HIGH: self.angle_high,
+            ATTR_FLOW_DIRECTION_FRONT: self.flow_direction_front,
+            ATTR_TIMER: self.timer,
```

Comment:
What kind of time is this? We only allow absolute utc time in state attributes.

Reply:
When will the integer timer value be updated? How often? Is it a relative time? We're not allowed to store relative times in the state attributes. That would spam the state machine.

Reply:
The fan sends an update message each time its state changes (some setting is set), that contains all attributes(power settings, filter state, speed) including the sleep timer that looks like this:

`fan_power=ON,front_direction=ON,auto_mode=ON,oscillation_status=ON,oscillation=ON,night_mode=ON,continuous_monitoring=ON,fan_state=FAN,night_mode_speed=0004,speed=AUTO,carbon_filter_state=0085,hepa_filter_state=0085,sleep_timer=OFF,oscillation_angle_low=0005,oscillation_angle_high=0185`

I just double checked it and the state message is sent only when its internal state changes. So for example if I set the timer to 15 minutes I'll immediately get a state message back that will look like this: 

`fan_power=ON,front_direction=ON,auto_mode=ON,oscillation_status=ON,oscillation=ON,night_mode=ON,continuous_monitoring=ON,fan_state=FAN,night_mode_speed=0004,speed=AUTO,carbon_filter_state=0085,hepa_filter_state=0085,sleep_timer=0015,oscillation_angle_low=0005,oscillation_angle_high=0185`

Then assuming that I don't issue any other commands to the fan, I'll receive the next state message after 15 minutes telling me that the fan turned itself off that will look like this:

`fan_power=OFF,front_direction=ON,auto_mode=ON,oscillation_status=ON,oscillation=ON,night_mode=ON,continuous_monitoring=ON,fan_state=OFF,night_mode_speed=0004,speed=AUTO,carbon_filter_state=0085,hepa_filter_state=0085,sleep_timer=OFF,oscillation_angle_low=0005,oscillation_angle_high=0185
`

Reply:
Good! Since polling is off, no unnecessary updates will occur.